### PR TITLE
Ian Lyell final PR

### PR DESCRIPTION
**Pull request by ILyell**

### CODE REVIEW

File: lib/gradebook.rb [hunk 1]
```diff
@@ -1,3 +1,46 @@
...
+
+    def add_course(course)
+        @courses << course
+        @courses
+    end
+
+    def list_all_students
+        student_list = {}
+        @courses.each do |course|
+            student_list[course.name] = course.students
```
Comment (KatBrandt) [inline lib/gradebook.rb:18]: Remember, we wanted the keys to be Course Objects, not just the name of the course. So just remove the `.name` there after course and this is exactly it! Great job with this.

File: lib/gradebook.rb [hunk 2]
```diff
@@ -1,3 +1,46 @@
...
+        end
+        student_list
+    end
+    
+    def students_below(score)
+        list_all_students.values.flatten.find_all do |student|
+            student.grade < score
+        end
```
Comment (KatBrandt) [inline lib/gradebook.rb:26]: SWEET!!!!!! Love your logic here.

File: lib/gradebook.rb [hunk 3]
```diff
@@ -1,3 +1,46 @@
...
+    end
+    
+    def all_grades
+        grades_list = {}
+        @courses.each do |course|
+            grade_array = []
+            course.students.each {|student| grade_array << student.grade}
+            grades_list[course.name] = grade_array
+        end
+        grades_list
```
Comment (KatBrandt) [inline lib/gradebook.rb:36]: yup yup!

File: lib/gradebook.rb [hunk 4]
```diff
@@ -1,3 +1,46 @@
...
+    end
+
+    def students_in_range(min, max)
+        list_all_students.values.flatten.find_all do |student|
+            if (student.grade > min) && (student.grade < max)
+                student
+            end
```
Comment (KatBrandt) [inline lib/gradebook.rb:43]: So since you're using find_all, you don't need the if stmt, just the conditions. Great job with these methods!

File: spec/gradebook_spec.rb [hunk 5]
```diff
@@ -0,0 +1,165 @@
+require 'rspec'
+require './lib/course'
+require './lib/student'
+require './lib/gradebook'
+
+RSpec.describe Gradebook do
+    describe '#initialize' do
+        it 'exist, and takes a string argument' do
+            gradebook = Gradebook.new("Mr.Dr.Prof. Patrick")
```
Comment (KatBrandt) [inline spec/gradebook_spec.rb:9]: haha, love the instructor name.

File: spec/gradebook_spec.rb [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            student2.log_score(45)
+
+            course_1.enroll(student1)
+            course_1.enroll(student2)
+
+            gradebook.add_course(course_1)
+
+            expect(gradebook.students_below(80)).to eq([student2])
+            expect(gradebook.students_below(90)).to eq([student1, student2])
+            expect(gradebook.students_below(40)).to eq([])
```
Comment (KatBrandt) [inline spec/gradebook_spec.rb:93]: Great job testing multiple thresholds here! Really makes the test more robust.

File: spec/gradebook_spec.rb [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            student4.log_score(70)
+            student4.log_score(65)
+            course_1.enroll(student1)
+            course_1.enroll(student2)
+            course_2.enroll(student3)
+            course_2.enroll(student4)
+            gradebook.add_course(course_1)
+            gradebook.add_course(course_2)
+            
+            expect(gradebook.all_grades).to eq({"Calculus" => [student1.grade, student2.grade], "English"=> [student3.grade, student4.grade]})
```
Comment (KatBrandt) [inline spec/gradebook_spec.rb:125]: Same thing here with the keys, we want them to be Objects instead of just the name. Objects are usually preferred because they are rich with info as opposed to just a string. Great job though!

File: spec/gradebook_spec.rb [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            #average = 67.5
+            course_1.enroll(student1)
+            course_1.enroll(student2)
+            course_2.enroll(student3)
+            course_2.enroll(student4)
+            gradebook.add_course(course_1)
+            gradebook.add_course(course_2)
+
+            expect(gradebook.students_in_range(60, 80)).to eq([student3, student4])
+            expect(gradebook.students_in_range(40, 80)).to eq([student2, student3, student4])
```
Comment (KatBrandt) [inline spec/gradebook_spec.rb:161]: Awesome job. Your tests are really robust. You can start thinking more about edge case testing. For example, in the Student#grade method, what if a student had no scores? Would the method break? etc.
